File: egottol/engines/ai/nsp.py
"""Neural Signal Processor — implements NSP_AI component behavior."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Tuple

import numpy as np


@dataclass
class NSPConfig:
    moving_avg_window: int = 5
    spectral_gate_threshold: float = 0.15
    fft_bins: int = 32
    anomaly_z_threshold: float = 3.0


class NeuralSignalProcessor:
    """AI-driven signal cleanup, classification, and anomaly detection."""

    def __init__(self, config: Optional[NSPConfig] = None):
        self.config = config or NSPConfig()
        self._classifier_weights: Optional[np.ndarray] = None
        self._classifier_bias: Optional[np.ndarray] = None
        self._feature_mean: Optional[np.ndarray] = None
        self._feature_std: Optional[np.ndarray] = None

    def denoise(self, signal: np.ndarray, sample_rate: float = 1.0) -> np.ndarray:
        """Moving average followed by spectral gating."""
        x = np.asarray(signal, dtype=float).reshape(-1)
        if x.size == 0:
            return x

        window = max(int(self.config.moving_avg_window), 1)
        kernel = np.ones(window, dtype=float) / window
        smoothed = np.convolve(x, kernel, mode="same")

        n = smoothed.size
        spectrum = np.fft.rfft(smoothed)
        freqs = np.fft.rfftfreq(n, d=1.0 / max(sample_rate, 1e-9))
        magnitude = np.abs(spectrum)
        peak = np.max(magnitude) if magnitude.size else 1.0
        gate = magnitude >= self.config.spectral_gate_threshold * peak
        filtered = np.fft.irfft(spectrum * gate, n=n)
        return filtered

    def _fft_features(self, signal: np.ndarray) -> np.ndarray:
        x = np.asarray(signal, dtype=float).reshape(-1)
        spectrum = np.abs(np.fft.rfft(x))
        bins = self.config.fft_bins
        if spectrum.size >= bins:
            idx = np.linspace(0, spectrum.size - 1, bins, dtype=int)
            return spectrum[idx]
        out = np.zeros(bins, dtype=float)
        out[: spectrum.size] = spectrum
        return out

    def train_classifier(
        self,
        signals: np.ndarray,
        labels: np.ndarray,
    ) -> None:
        """Fit a linear classifier on FFT features (least squares)."""
        signals = np.atleast_2d(signals)
        labels = np.asarray(labels, dtype=int).reshape(-1)
        features = np.array([self._fft_features(s) for s in signals], dtype=float)
        self._feature_mean = features.mean(axis=0)
        self._feature_std = np.maximum(features.std(axis=0), 1e-9)
        normed = (features - self._feature_mean) / self._feature_std

        classes = np.unique(labels)
        n_classes = classes.size
        n_features = normed.shape[1]
        y_onehot = np.zeros((labels.size, n_classes), dtype=float)
        for i, c in enumerate(classes):
            y_onehot[labels == c, i] = 1.0

        aug = np.hstack([normed, np.ones((normed.shape[0], 1))])
        w_aug, _, _, _ = np.linalg.lstsq(aug, y_onehot, rcond=None)
        self._classifier_weights = w_aug[:n_features, :].T
        self._classifier_bias = w_aug[n_features, :]
        self._class_labels = classes

    def classify(self, signal: np.ndarray) -> Tuple[int, np.ndarray]:
        """Classify a signal using FFT features and the linear readout."""
        if self._classifier_weights is None:
            self.train_classifier(
                np.array([signal, signal * 0.5]),
                np.array([0, 1]),
            )

        feat = self._fft_features(signal)
        feat = (feat - self._feature_mean) / self._feature_std
        logits = self._classifier_weights @ feat + self._classifier_bias
        exp = np.exp(logits - np.max(logits))
        probs = exp / np.sum(exp)
        idx = int(np.argmax(probs))
        return int(self._class_labels[idx]), probs

    def anomaly_detect(
        self,
        signal: np.ndarray,
        reference: Optional[np.ndarray] = None,
    ) -> Tuple[bool, float]:
        """Z-score anomaly test against reference or running mean of signal."""
        x = np.asarray(signal, dtype=float).reshape(-1)
        ref = x if reference is None else np.asarray(reference, dtype=float).reshape(-1)
        mean = float(np.mean(ref))
        std = float(np.std(ref))
        if std < 1e-12:
            std = 1e-12
        z = float(np.max(np.abs(x - mean)) / std)
        return z > self.config.anomaly_z_threshold, z

File: core/ai/nsp.cpp
#include "nsp.h"

#include <algorithm>
#include <cmath>
#include <complex>
#include <numeric>
#include <stdexcept>

#include "../matrix.h"
#include "fft.h"

namespace deepiri {

namespace {

// numpy's np.convolve(x, kernel, mode="same") for kernel.size() <= x.size():
// take the central `x.size()` samples of the full convolution.
std::vector<double> convolveSame(const std::vector<double>& x, const std::vector<double>& kernel) {
    const size_t m = x.size();
    const size_t k = kernel.size();
    std::vector<double> full(m + k - 1, 0.0);
    for (size_t i = 0; i < m; ++i) {
        for (size_t j = 0; j < k; ++j) {
            full[i + j] += x[i] * kernel[j];
        }
    }
    const size_t offset = (k - 1) / 2;
    return std::vector<double>(full.begin() + static_cast<long>(offset),
                                full.begin() + static_cast<long>(offset + m));
}

// fft.h only supports power-of-two sizes, whereas numpy's rfft/irfft handle
// any length; signals are zero-padded up to the next power of two here and
// truncated back afterward. This changes the effective frequency bin
// spacing versus numpy for non-power-of-two inputs — an intentional
// deviation flagged in the port notes rather than adding a mixed-radix FFT.
std::vector<std::complex<double>> paddedForwardFFT(const std::vector<double>& x, size_t& npad) {
    npad = nextPowerOfTwo(std::max<size_t>(x.size(), 1));
    std::vector<std::complex<double>> padded(npad, std::complex<double>(0.0, 0.0));
    for (size_t i = 0; i < x.size(); ++i) padded[i] = std::complex<double>(x[i], 0.0);
    return fft(std::move(padded));
}

size_t linspaceIndex(size_t i, size_t bins, size_t spectrumSize) {
    if (bins <= 1 || spectrumSize <= 1) return 0;
    double value = static_cast<double>(i) * static_cast<double>(spectrumSize - 1) /
                   static_cast<double>(bins - 1);
    return static_cast<size_t>(value);  // truncation matches numpy's astype(int)
}

}  // namespace

NSP::NSP(const NSPConfig& config) : config_(config) {}

std::vector<double> NSP::denoise(const std::vector<double>& signal, double sampleRate) const {
    (void)sampleRate;
    if (signal.empty()) return signal;

    const size_t window = std::max<size_t>(static_cast<size_t>(config_.moving_avg_window), 1);
    std::vector<double> kernel(window, 1.0 / static_cast<double>(window));
    std::vector<double> smoothed = convolveSame(signal, kernel);

    const size_t n = smoothed.size();
    size_t npad = 0;
    std::vector<std::complex<double>> spectrum = paddedForwardFFT(smoothed, npad);

    double peak = 0.0;
    for (const auto& c : spectrum) peak = std::max(peak, std::abs(c));
    if (peak <= 0.0) peak = 1.0;

    std::vector<std::complex<double>> gated(npad);
    for (size_t i = 0; i < npad; ++i) {
        bool pass = std::abs(spectrum[i]) >= config_.spectral_gate_threshold * peak;
        gated[i] = pass ? spectrum[i] : std::complex<double>(0.0, 0.0);
    }

    std::vector<std::complex<double>> restored = ifft(std::move(gated));
    std::vector<double> filtered(n);
    for (size_t i = 0; i < n; ++i) filtered[i] = restored[i].real();
    return filtered;
}

std::vector<double> NSP::fftFeatures(const std::vector<double>& signal) const {
    size_t npad = 0;
    std::vector<std::complex<double>> spectrum = paddedForwardFFT(signal, npad);
    const size_t spectrumSize = npad / 2 + 1;

    const size_t bins = config_.fft_bins;
    std::vector<double> out(bins, 0.0);
    if (spectrumSize >= bins) {
        for (size_t i = 0; i < bins; ++i) {
            out[i] = std::abs(spectrum[linspaceIndex(i, bins, spectrumSize)]);
        }
    } else {
        for (size_t i = 0; i < spectrumSize; ++i) out[i] = std::abs(spectrum[i]);
    }
    return out;
}

void NSP::trainClassifier(const std::vector<std::vector<double>>& signals, const std::vector<int>& labels) {
    const size_t nSamples = signals.size();
    const size_t bins = config_.fft_bins;

    std::vector<std::vector<double>> features(nSamples);
    for (size_t s = 0; s < nSamples; ++s) features[s] = fftFeatures(signals[s]);

    featureMean_.assign(bins, 0.0);
    featureStd_.assign(bins, 0.0);
    for (size_t j = 0; j < bins; ++j) {
        double sum = 0.0;
        for (size_t s = 0; s < nSamples; ++s) sum += features[s][j];
        featureMean_[j] = sum / static_cast<double>(nSamples);
    }
    for (size_t j = 0; j < bins; ++j) {
        double sumSq = 0.0;
        for (size_t s = 0; s < nSamples; ++s) {
            double d = features[s][j] - featureMean_[j];
            sumSq += d * d;
        }
        featureStd_[j] = std::max(std::sqrt(sumSq / static_cast<double>(nSamples)), 1e-9);
    }

    std::vector<int> classes = labels;
    std::sort(classes.begin(), classes.end());
    classes.erase(std::unique(classes.begin(), classes.end()), classes.end());
    classLabels_ = classes;
    const size_t nClasses = classes.size();

    std::vector<std::vector<double>> yOneHot(nSamples, std::vector<double>(nClasses, 0.0));
    for (size_t s = 0; s < nSamples; ++s) {
        auto it = std::find(classes.begin(), classes.end(), labels[s]);
        yOneHot[s][static_cast<size_t>(it - classes.begin())] = 1.0;
    }

    const size_t nAug = bins + 1;
    Matrix aug(nSamples, nAug);
    for (size_t s = 0; s < nSamples; ++s) {
        for (size_t j = 0; j < bins; ++j) {
            aug.at(s, j) = (features[s][j] - featureMean_[j]) / featureStd_[j];
        }
        aug.at(s, bins) = 1.0;
    }

    Matrix augT = aug.transpose();
    Matrix ata = augT * aug;
    Matrix aty = augT * Matrix(yOneHot);

    // matrix.h has no SVD/pseudoinverse, so unlike numpy's np.linalg.lstsq
    // (which is rank-deficiency-safe), the normal equations here need a
    // tiny Tikhonov nudge to stay solvable when samples < features + 1.
    for (size_t i = 0; i < nAug; ++i) ata.at(i, i) += 1e-8;

    Matrix wAug(nAug, nClasses, 0.0);
    for (size_t c = 0; c < nClasses; ++c) {
        std::vector<double> col(nAug);
        for (size_t i = 0; i < nAug; ++i) col[i] = aty.at(i, c);
        std::vector<double> w = ata.solveGaussian(col);
        for (size_t i = 0; i < nAug; ++i) wAug.at(i, c) = w[i];
    }

    classifierWeights_.assign(nClasses, std::vector<double>(bins, 0.0));
    classifierBias_.assign(nClasses, 0.0);
    for (size_t c = 0; c < nClasses; ++c) {
        for (size_t j = 0; j < bins; ++j) classifierWeights_[c][j] = wAug.at(j, c);
        classifierBias_[c] = wAug.at(bins, c);
    }
    classifierTrained_ = true;
}

std::pair<int, std::vector<double>> NSP::classify(const std::vector<double>& signal) {
    if (!classifierTrained_) {
        std::vector<double> halved(signal.size());
        for (size_t i = 0; i < signal.size(); ++i) halved[i] = signal[i] * 0.5;
        trainClassifier({signal, halved}, {0, 1});
    }

    std::vector<double> feat = fftFeatures(signal);
    for (size_t j = 0; j < feat.size(); ++j) feat[j] = (feat[j] - featureMean_[j]) / featureStd_[j];

    const size_t nClasses = classifierWeights_.size();
    std::vector<double> logits(nClasses, 0.0);
    for (size_t c = 0; c < nClasses; ++c) {
        double sum = classifierBias_[c];
        for (size_t j = 0; j < feat.size(); ++j) sum += classifierWeights_[c][j] * feat[j];
        logits[c] = sum;
    }

    double maxLogit = *std::max_element(logits.begin(), logits.end());
    std::vector<double> probs(nClasses);
    double sumExp = 0.0;
    for (size_t c = 0; c < nClasses; ++c) {
        probs[c] = std::exp(logits[c] - maxLogit);
        sumExp += probs[c];
    }
    for (double& p : probs) p /= sumExp;

    size_t argmax = static_cast<size_t>(std::max_element(probs.begin(), probs.end()) - probs.begin());
    return {classLabels_[argmax], probs};
}

std::pair<bool, double> NSP::anomalyDetect(const std::vector<double>& signal,
                                            const std::vector<double>* reference) const {
    const std::vector<double>& ref = reference ? *reference : signal;
    double mean = std::accumulate(ref.begin(), ref.end(), 0.0) / static_cast<double>(ref.size());
    double sumSq = 0.0;
    for (double v : ref) sumSq += (v - mean) * (v - mean);
    double stddev = std::sqrt(sumSq / static_cast<double>(ref.size()));
    if (stddev < 1e-12) stddev = 1e-12;

    double maxAbsDev = 0.0;
    for (double v : signal) maxAbsDev = std::max(maxAbsDev, std::abs(v - mean));
    double z = maxAbsDev / stddev;
    return {z > config_.anomaly_z_threshold, z};
}

}  // namespace deepiri
